R/di-fit.R

```diff
@@ -0,0 +1,276 @@
+        if (row_n != 1) {
+          i <- seq_len(row_n)
+          j <- row_n
+          lower_distances <- distances[nrow_training*(i-1) - i*(i-1)/2 + j-i]
+        }
+        if (row_n != nrow_training) {
+          i <- row_n
+          j <- seq.int(row_n + 1, nrow_training, 1)
+          higher_distances <- distances[nrow_training*(i-1) - i*(i-1)/2 + j-i]
+        }
+        min(c(lower_distances, higher_distances))
+      }
+    )
+  } else {
+    nrow_validation <- nrow(validation)
+    nrow_total <- nrow_validation + nrow_training
+    purrr::map_dbl(
+      seq_len(nrow_validation) + nrow_training,
+      function(row_n) {
+        i <- seq_len(nrow_training)
+        j <- row_n
+        min(distances[nrow_total*(i-1) - i*(i-1)/2 + j-i])
+      }
+    )
+  }
+}
+
+# -----------------------------------------------------------------------------
+# ------------------------ Model function bridge ------------------------------
+# -----------------------------------------------------------------------------
+
+apd_di_bridge <- function(training, validation, importance, ...) {
+  blueprint <- training$blueprint
+  training <- training$predictors
+  all_numeric <- purrr::map_lgl(training, is.numeric)
+
+  if (!is.null(validation)) {
+    validation <- validation$predictors
+    all_numeric <- c(all_numeric, purrr::map_lgl(validation, is.numeric))
+  }
+
+  all_numeric <- all(all_numeric)
+
+  if (!all_numeric) {
+    rlang::abort(
+      "All predictors must be numeric",
+      call = rlang::caller_env()
+    )
+  }
+
+  all_importance <- all(names(training) %in% importance[["Variable"]])
+
+  if (!all_importance) {
+    rlang::abort(
+      "All predictors must have an importance value in `importance`",
+      call = rlang::caller_env()
+    )
+  }
+
+  fit <- apd_di_impl(training, validation, importance, ...)
+
+  new_apd_di(
+    training = fit$training,
+    importance = fit$importance,
+    sds = fit$sds,
+    means = fit$means,
+    d_bar = fit$d_bar,
+    aoa_threshold = fit$aoa_threshold,
+    blueprint = blueprint
+  )
+}
+
+# -----------------------------------------------------------------------------
+# ----------------------- Model function interface ----------------------------
+# -----------------------------------------------------------------------------
+
+#' Fit a `apd_di`
+#'
+#' `apd_di()` fits a model.
+#'
+#' @param x,y Depending on the context:
+#'
+#'   * A __data frame__ of predictors.
+#'   * A __matrix__ of predictors.
+#'   * A __recipe__ specifying a set of preprocessing steps
+#'     created from [recipes::recipe()].
+#'
+#'  If `y` is `NULL`, then this function will use within-sample distances.
+#'  This may result in too large an area of applicability being calculated.
+#'
+#' @param data,validation When a __recipe__ or __formula__ is used, `data` is specified as:
+#'
+#'   * A __data frame__ containing the predictors.
+#'
+#'  If `validation` is `NULL`, then this function will use within-sample distances.
```

Comment:
Can you give information here of what is being validated? Are there specific metrics being computed? What is the downside of not using a validation set etc.

Reply:
I tried to clarify this in 86b1a5de37230c2aca4daef69ae3de936f02dc79 and 33faf2e but am not sure if I quite nailed it -- it should be the same hold-out set as you used in calculating all other model metrics. Please let me know if there's still anything unclear!